C# v7.x: Local functions

RexJaeschke opened the pull request.

RexJaeschke committed:
Update statements.md

RexJaeschke committed:
Update unsafe-code.md

RexJaeschke committed:
Add semantics and example for local functions v7.0 feature

RexJaeschke committed:
removed local function unsafe grammar extension

With the consolidation of the unsafe grammar in PR #233, this text is no longer needed.

KalleOlaviNiemitalo commented on standard/statements.md:
```diff
@@ -352,6 +355,63 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+It is a compile-time error for a local function to declare a parameter or local variable with the same name as one declared in the enclosing scope.
+
+Local function declaration statements are always reachable.
```
```C#
class C {
    int F() {
        if (true) {
            void G() {}
            return 1;
        } else {
            void H() {}
            return 2;
        }
    }
}
```

Is there a rule that says `return 1;` is reachable and `return 2;` is not? According to [13.3.2 Statement lists](https://github.com/dotnet/csharpstandard/blob/c86c9b0f464572b060dfa6688968237a4d23ad9e/standard/statements.md#1332-statement-lists), it depends on whether "The end point of the preceding statement is reachable." But if both local function declaration statements are reachable, then I don't see why the end point of `void G() {}` would be reachable and that of `void H() {}` would not.

BillWagner replied:
There isn't such a rule right now. The compiler determines that H() is reachable.

BillWagner replied:
Discussed in the committee meeting:

```suggestion
Local function bodies are always reachable. The endpoint of a local function is reachable if the beginning point of the local function is reachable.
```

RexJaeschke committed:
include support for local functions

RexJaeschke committed:
add support for local functions

KalleOlaviNiemitalo commented on standard/basic-concepts.md:
```diff
@@ -62,8 +62,8 @@ There are several different types of declaration spaces, as described in the fol
 -   Each enumeration declaration creates a new declaration space. Names are introduced into this declaration space through *enum_member_declarations*.
--   Each method declaration, property declaration, property accessor declaration, indexer declaration, indexer accessor declaration, operator declaration, instance constructor declaration and anonymous function creates a new declaration space called a ***local variable declaration space***. Names are introduced into this declaration space through formal parameters (*fixed_parameter*s and *paramet ...
--   Each *block* or *switch_block*, as well as a `for`, `foreach`, and `using` statement, creates a local variable declaration space for local variables and local constants. Names are introduced into this declaration space through *local_variable_declaration*s and *local_constant_declaration*s. Note that blocks that occur as or within the body of a function member or anonymous function are nested ...
+-   Each method declaration, property declaration, property accessor declaration, indexer declaration, indexer accessor declaration, operator declaration, instance constructor declaration, anonymous function, and local function creates a new declaration space called a ***local variable declaration space***. Names are introduced into this declaration space through formal parameters (*fixed_paramet ...
```
Perhaps this should say ``the `value` name`` rather than `the valuename`. That was not introduced in this PR though.

BillWagner committed:
fix build issues

BillWagner commented on standard/statements.md:
```diff
@@ -414,6 +417,65 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+Unless specified otherwise below, the semantics of all grammar elements is the same as for *method_declaration* ([§14.6.1](classes.md#1461-general), read in the context of a local function instead of a method.
+
+A *local_function_declaration* may include one `async` ([§14.15](classes.md#1415-async-functions)) modifier and one `unsafe` ([§22.1](unsafe-code.md#221-general)) modifier. If the declaration includes the `async` modifier then the return type shall be `void` or a task type  ([§14.15.1](classes.md#14151-general)).
```
This will likely need to be revisited in light of generalized async return types. (`ValueTask<T>` works, among others).

BillWagner commented on standard/statements.md:
```diff
@@ -414,6 +417,65 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+A *local_function_declaration* may include one `async` ([§14.15](classes.md#1415-async-functions)) modifier and one `unsafe` ([§22.1](unsafe-code.md#221-general)) modifier. If the declaration includes the `async` modifier then the return type shall be `void` or a task type  ([§14.15.1](classes.md#14151-general)).
+
+A local function is declared at block scope, and that function may capture variables from the enclosing scope. Each call of the function requires captured variables to be definitely assigned before the call. The compiler shall determine which variables are definitely assigned on return.
```
Adding @RikkiGibson

I found some interesting bits here, some of which don't apply until C# 8 (the reason I tagged Rikki).  Consider this example:

```csharp
public class C
{
    private string field;
    public C()
    {
        string local;
        F();
        String s = field;
        field = local;        
        void F()
        {
            this.field = "assigned";
            local = "assigned";
        }
    }
}
```

The compiler determines that the local function `F` does definitely assign `local` (`field` is definitely assigned to `null`). However, the nullable analysis doesn't set the null state of `field` or `local` to *not null* after `F()` is called. That makes sense the nullable reference spec doesn't specify that the compiler determines the null state of variables when a local function is called, but it seems odd to me. (That question doesn't need to be answered until we get to C# 8, so not blocking this PR).

RikkiGibson replied:
It is definitely a strange corner of our flow analysis. We made a deliberate decision to *not* "flow out" nullable state changes from local functions. They only "flow in"--i.e. when you have multiple call sites, the compiler determines a "merged worst state" to use as the initial state to analyze the local function. This was due to the performance cost and complexity of definite assignment analysis versus nullable analysis.

BillWagner reviewed (commented):
This looks great @RexJaeschke  Should we mark this ready for review?

I've got one note for how local functions interacts with generalized async return types. 

I've also got one second question that won't need to be addressed until C# 8.

BillWagner commented on standard/statements.md:
```diff
@@ -414,6 +417,65 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+A *local_function_declaration* may include one `async` ([§14.15](classes.md#1415-async-functions)) modifier and one `unsafe` ([§22.1](unsafe-code.md#221-general)) modifier. If the declaration includes the `async` modifier then the return type shall be `void` or a task type  ([§14.15.1](classes.md#14151-general)).
+
+A local function is declared at block scope, and that function may capture variables from the enclosing scope. Each call of the function requires captured variables to be definitely assigned before the call. The compiler shall determine which variables are definitely assigned on return.
```
The statement "Each call of the function requires captured variables to be definitely assigned before the call." is incorrect. Captured variables must only be definitely assigned before it is *read* by the body of the local function.

BillWagner replied:
Add informative examples here as well.

BillWagner reviewed (commented):
Add notes from committee meeting.

BillWagner committed:
Apply suggestions from code review

Co-authored-by: KalleOlaviNiemitalo <kon@iki.fi>

BillWagner committed:
clarify definite assignment, add examples

Addresses comments in committee meeting:

- https://github.com/dotnet/csharpstandard/pull/104#discussion_r870744332
- https://github.com/dotnet/csharpstandard/pull/104#discussion_r870745137

BillWagner commented:
This is now ready for final review.

@jskeet Should I tag others for review?

gafter commented:
A lot of the work on this feature was done by @agocke.  It would be good to get his feedback.

agocke commented:
I'll try to take a look. It looks like the definite assignment section needs work. Unlike lambdas, local functions use a more precise definite assignment when not converted to delegates. I'll try to write up some appropriate spec language for this. Overview of the differences is at https://commentout.com/localfunc.html

RexJaeschke commented:
@BillWagner I think the header and modifiers rules needs tweaking:

We have

```ANTLR
local_function_header
    : local_function_modifiers? return_type identifier type_parameter_list?
        ( formal_parameter_list? ) type_parameter_constraints_clause*
    ;
local_function_modifiers
    : 'async'
    | 'unsafe'
    ;
```

That allows a local function to have no modifiers, `async`, or `unsafe`, but it does *not* allow `async unsafe`, which I think is a valid combination. [In V8, we'll add `static`, and in V9, `extern`.]

Following the corresponding pattern for modifiers on methods ([14.16.1](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/classes.md#1461-general)), the rule should be renamed `local_function_modifier` (singular), with 0 or more occurrences allowed, with both rules changed, as follows:

```ANTLR
local_function_header
    : local_function_modifier* return_type identifier type_parameter_list?
        ( formal_parameter_list? ) type_parameter_constraints_clause*
    ;
local_function_modifier
    : 'async'
    | 'unsafe'
    ;
```

RexJaeschke commented:
Another grammar problem!

Not surprisingly, the header rule uses the same rules `type_parameter_list` and `formal_parameter_list` as `method_header` in [14.16.1](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/classes.md#1461-general), as follows:. 

```ANTLR
local_function_header
    : local_function_modifiers? return_type identifier type_parameter_list?
        ( formal_parameter_list? ) type_parameter_constraints_clause*
    ;
```

However, these two rules **allow** attributes, yet it isn't until V9 that support for attributes is added to type parameters and formal parameters in local functions (as well as to local functions themselves)). So, for V7 and V8, we need to come up with a stripped-down set of these two rules that *disallow* attributes.

Here's my proposal for that (with admittedly long, but clear rule names):

```ANTLR
local_function_header
    : local_function_modifier* return_type identifier type_parameter_list_with_no_attributes?
        ( formal_parameter_list_with_no_attributes? ) type_parameter_constraints_clause*
    ;

local_function_modifier
    : 'async'
    | 'unsafe'
    ;

type_parameter_list_with_no_attributes
    : '<' type_parameters_with_no_attributes '>'
    ;

type_parameters_with_no_attributes
    : type_parameter
    | type_parameters_with_no_attributes ',' type_parameter
    ;

formal_parameter_list_with_no_attributes
    : fixed_parameters_with_no_attributes
    | fixed_parameters_with_no_attributes ',' parameter_array_with_no_attributes
    | parameter_array_with_no_attributes
    ;

fixed_parameters_with_no_attributes
    : fixed_parameter_with_no_attributes (',' fixed_parameter_with_no_attributes)*
    ;

fixed_parameter_with_no_attributes
    : parameter_modifier? type identifier default_argument?
    ;

parameter_array_with_no_attributes
    : 'params' array_type identifier
    ;
```

**This grammar includes the suggestion from my [earlier comment](https://github.com/dotnet/csharpstandard/pull/104#issuecomment-1126378608).**

gafter commented:
Rather than modifying the syntax, I suggest adding a semantic rule that attributes are not permitted on local functions.  Then, later, we can just remove that one sentence.

RexJaeschke commented:
Yes, @gafter, you are quite right; adding a constraint is a far less intrusive way of dealing with this.

I suggest that after the following text

> A *local_function_declaration* may include one `async` ([§14.15](classes.md#1415-async-functions)) modifier and one `unsafe` ([§22.1](unsafe-code.md#221-general)) modifier. If the declaration includes the `async` modifier then the return type shall be `void` or a task type  ([§14.15.1](classes.md#14151-general)).

we add

>  It is a compile-time error for *type_parameter_list* or *formal_parameter_list* to contain *attributes*.

However, we'll still need the grammar change w.r.t the modifiers, as I mentioned earlier.

RexJaeschke commented:
Another grammar issue. We have the following

```ANTLR
local_function_body
    : block
    | '=>' expression ';'
    ;
```

It is likely that this was originally cloned from *method_body* ([14.6.1](standard/classes.md#1461-general), with the ` | ';'` rule omitted, as external local-function support wasn't added until V9). However, in V6, we added another rule to *method_body*, as follows:

```ANTLR
    | '=>' null_conditional_invocation_expression ';'
```

Should that rule also be added to *local_function_body*? If so, then Nigel's corresponding grammar note from the discussion of *method_body* should be adapted for here, as follows:

> Grammar note: When recognising a *local_function_body* if both the *null_conditional_invocation_expression* and *expression* alternatives are applicable then the former shall be chosen.

jskeet commented on standard/statements.md:
```diff
@@ -414,6 +417,100 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+A local function is declared at block scope, and that function may capture variables from the enclosing scope. Captured variables must be definitely assigned before read by the body of the local function in each call to the function. The compiler shall determine which variables are definitely assigned on return.
+
+> *Example*: The following example demonstrates definite assignment for captured variables in local functions. If a local function reads a captured variable before writing it, the captured variable must be definitely assigned before calling the local function. The local function `F1` reads `s` without assigning it. It is an error if `F1` is called before `s` is definitely assigned. `F2` assigns ` ...
```
This part:

> Furthermore, `F1` may be called after `F2` because `i` is definitely assigned in `F2`.

... doesn't make much sense as `F1` doesn't read `i`. (I suspect at some point both functions used the same variable.)

BillWagner replied:
updated.

jskeet commented on standard/statements.md:
```diff
@@ -414,6 +417,100 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+A local function may be called from a lexical point prior to its definition. However, it is a compile-time error for the function to be declared lexically prior to the declaration of a variable it wishes to capture.
+
+It is a compile-time error for a local function to declare a parameter or local variable with the same name as one declared in the enclosing scope.
```
I don't think this restriction is actually correct... or at least it isn't now. The code below is valid:

```csharp
using System;

class Program
{
    static void Main()
    {
        int x = 10;
        Local1("hello");
        Local2();
        
        void Local1(string x)
        {
            Console.WriteLine(x);
        }
        
        void Local2()
        {
            string x = "hello";
            Console.WriteLine(x);
        }
    }
}
```

Is this something that changed after C# 7?

BillWagner replied:
Yes. I just verified this. Your example doesn't compiler in C# 7.3, but does in 8.0. I wrote #590 to capture this.

jskeet replied:
Ha - looks like I've noticed this a second time above...

jskeet commented on standard/statements.md:
````diff
@@ -414,6 +417,100 @@ The scope of a local constant is the block in which the declaration occurs. It i
+> ```csharp
+> public static IEnumerable<char> AlphabetSubset(char start, char end)
+> {
+>     if (start < 'a' || start > 'z')
````
Let's add braces to this example, and a black line after the return statement.

jskeet commented on standard/statements.md:
```diff
@@ -414,6 +417,100 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+Local function bodies are always reachable. The endpoint of a local function is reachable if the beginning point of the local function is reachable.
+
+> *Example*: There are two common use cases for local functions: public iterator methods and public async methods. In iterator methods, any exceptions are observed only when calling code that enumerates the returned sequence. In async methods, any exceptions are only observed when the returned Task is awaited. The following example demonstrates separating parameter validation from the iterator im ...
```
Normally "example" refers to the preceding text - this is more of a "general use case" - in particular it's not about reachability. Maybe the example should be right at the start of the section, under the first sentence of description?

BillWagner replied:
Added brackets. I think I moved it to the right location (after the grammar).

BillWagner committed:
respond to feedback.

BillWagner commented:
> Another grammar issue. We have the following
> 
> ```antlr
> local_function_body
>     : block
>     | '=>' expression ';'
>     ;
> ```
> 
> It is likely that this was originally cloned from _method_body_ ([14.6.1](standard/classes.md#1461-general), with the ` | ';'` rule omitted, as external local-function support wasn't added until V9). However, in V6, we added another rule to _method_body_, as follows:
> 
> ```antlr
>     | '=>' null_conditional_invocation_expression ';'
> ```
> 
> Should that rule also be added to _local_function_body_? If so, then Nigel's corresponding grammar note from the discussion of _method_body_ should be adapted for here, as follows:
> 
> > Grammar note: When recognising a _local_function_body_ if both the _null_conditional_invocation_expression_ and _expression_ alternatives are applicable then the former shall be chosen.

I verified that yes, the `null_conditional_invocation_expression` is a valid expression for a local function. I made these edits.

BillWagner committed:
grammar updates

BillWagner commented:
I've updated this based on all the feedback *except* the definite assignment question.

Should we make a separate PR that updates [9.4](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/variables.md#94-definite-assignment) by adding a clause for assignment in local functions? Or should we put those updates in the statements section?

/cc @agocke

agocke commented:
Let me try to write down some simplified wording and see if you can massage it into spec language.

Local functions are analyzed in the context of their parent method. There are two new control flow paths that matter for local functions: calls and delegate conversions. Definite assignment for the body of each local function is defined separately for each call site. At each invocation, variables captured by the local function are considered definitely assigned if they were definitely assigned at the point of call. A control flow path also exists to the local function body at this point and is considered reachable. After a call to the local function, captured variables that were definitely assigned at every control point leaving the function (`return` statements, `yield` statements, `await` statements) are considered definitely assigned after the call location.

Delegate conversions have a control flow path to the local function body and captured variables are also considered assigned for the body if they are assigned before the conversion, but variables assigned by the local function are not considered assigned after the conversion.

Note: the above implies that bodies are re-analyzed for definite assignment at every local function invocation or delegate invocation. That's one way to achieve the required result, but for performance the compiler doesn't actually do so. The implementation should be equivalent, however.

BillWagner commented:
Meeting discussion item:

I can take Andy's text above and apply it to the standard. I'll propose that the definite assignment rules should be a new sub-clause of [9.4.4](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/variables.md#944-precise-rules-for-determining-definite-assignment), "Precise rules for definite assignment". 

@jskeet Can we discuss at today's meeting?

BillWagner commented:
Discussed in meeting:

- Add one or two sub-clauses under 9.4.4. to describe the definite assignment rules for local functions. Request another review after that's been added.

agocke commented:
Some other things about local functions that may be relevant:

- Unlike methods, they cannot be overloaded
- All calls to local functions must be static resolved. That means that the following is illegal:
```C#
using System;
public class C {
    public void M() {
        dynamic d = null;
        Local(d);
        void Local(params int[] xs) { }
    }
}
```
- `unsafe` uses the parent lexical scope, so `unsafe` on a parent method propagates to the nested local function
- `async` does *not* use the parent lexical scope, so the following is illegal
```C#
    public async Task M()
    {
        await Local();
        Task Local()
        {
            await Task.Delay(0);
        }
    }
```
- Local functions can be used to observe that the _scope_ and _lifetime_ of variables sometimes differ in C#. For example,
   ```C#
    public void M(Object o) {
        switch (o)
        {
            case 1:
                Local();
                break;
            case int x:
                void Local() { x = 1; }
                break;
        }
    }
   ```
   In this example the scope of the pattern variable `x` is the case clause it's in, while the lifetime is the whole switch statement

BillWagner commented:
Thanks @agocke 

Comments related to the edits I'm making now:

> Some other things about local functions that may be relevant:
> 
> * Unlike methods, they cannot be overloaded

Added to the new clause in statements.

> * All calls to local functions must be static resolved. That means that the following is illegal:
> 
> ```cs
> using System;
> public class C {
>     public void M() {
>         dynamic d = null;
>         Local(d);
>         void Local(params int[] xs) { }
>     }
> }
> ```

Added to clause in statements.

> * `unsafe` uses the parent lexical scope, so `unsafe` on a parent method propagates to the nested local function
> * `async` does _not_ use the parent lexical scope, so the following is illegal
> 
> ```cs
>     public async Task M()
>     {
>         await Local();
>         Task Local()
>         {
>             await Task.Delay(0);
>         }
>     }
> ```

Also added.

> * Local functions can be used to observe that the _scope_ and _lifetime_ of variables sometimes differ in C#. For example,
>   ```cs
>    public void M(Object o) {
>        switch (o)
>        {
>            case 1:
>                Local();
>                break;
>            case int x:
>                void Local() { x = 1; }
>                break;
>        }
>    }
>   ```
>       
>   In this example the scope of the pattern variable `x` is the case clause it's in, while the lifetime is the whole switch statement

I don't think this is needed in the standard. (It's interesting though.)

BillWagner committed:
Incorporate Andy's text

In this commit, create the new clause for the definite assignment rules for local functions. Move the existing example to that clause. Add text from Andy's comments and notes.

BillWagner committed:
finish local functions and definite assignment.

Incorporate the rules for definite assignment for captured variables in local functions. This is both for local function calls and for delegate conversion.

BillWagner commented:
@jskeet I've incorporated Andy's comments and description for definite assignment in local functions. I don't think there's enough time for everyone to review before our next meeting, but I'd like to put this one up for the following meeting.

BillWagner committed:
typo

Nigel-Ecma commented on standard/statements.md:
```diff
@@ -414,6 +417,78 @@ The scope of a local constant is the block in which the declaration occurs. It i
+>
+> *end example*
+
+Unless specified otherwise below, the semantics of all grammar elements is the same as for *method_declaration* ([§14.6.1](classes.md#1461-general), read in the context of a local function instead of a method.
```
Missing paren

```suggestion
Unless specified otherwise below, the semantics of all grammar elements is the same as for *method_declaration* ([§14.6.1](classes.md#1461-general)), read in the context of a local function instead of a method.
```

Nigel-Ecma commented on standard/variables.md:
```diff
@@ -802,6 +802,63 @@ For a *lambda_expression* or *anonymous_method_expression* *expr* with a body (e
+
+Delegate conversions have a control flow path to the local function body. Captured variables are definitely assigned for the body if they are definitely assigned before the conversion. Variables assigned by the local function are not considered assigned after the conversion.
+
+> *note:* the above implies that bodies are re-analyzed for definite assignment at every local function invocation or delegate conversion. Compilers are not required to re-analyze the body of a local function at each invocation or delegate conversion. The implementation must produce results equivalent to that description.
```
Typo:

```suggestion
> *Note:* the above implies that bodies are re-analyzed for definite assignment at every local function invocation or delegate conversion. Compilers are not required to re-analyze the body of a local function at each invocation or delegate conversion. The implementation must produce results equivalent to that description.
```

Nigel-Ecma commented on standard/statements.md:
```diff
@@ -414,6 +417,78 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+Grammar note: When recognising a *local_function_body* if both the *null_conditional_invocation_expression* and *expression* alternatives are applicable then the former shall be chosen.
+
+> *Example*: There are two common use cases for local functions: public iterator methods and public async methods. In iterator methods, any exceptions are observed only when calling code that enumerates the returned sequence. In async methods, any exceptions are only observed when the returned Task is awaited. The following example demonstrates separating parameter validation from the iterator im ...
```
Why *public* iterator/async, wouldn't the same benefits apply to those with other accessibilities, and I missing something?

BillWagner replied:
Good point. Let's remove "public":

```suggestion
> *Example*: There are two common use cases for local functions: iterator methods and async methods. In iterator methods, any exceptions are observed only when calling code that enumerates the returned sequence. In async methods, any exceptions are only observed when the returned Task is awaited. The following example demonstrates separating parameter validation from the iterator implementation using a local function:
```

Nigel-Ecma reviewed (commented):
A couple of typos and a question...

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Nigel-Ecma <perryresearch@zoot.net.nz>

BillWagner committed:
Apply suggestions from code review

jskeet commented on standard/basic-concepts.md:
```diff
@@ -64,8 +64,9 @@ There are several different types of declaration spaces, as described in the fol
 - Each enumeration declaration creates a new declaration space. Names are introduced into this declaration space through *enum_member_declarations*.
-- Each method declaration, property declaration, property accessor declaration, indexer declaration, indexer accessor declaration, operator declaration, instance constructor declaration and anonymous function creates a new declaration space called a ***local variable declaration space***. Names are introduced into this declaration space through formal parameters (*fixed_parameter*s and *parameter ...
-- Each *block* or *switch_block*, as well as a `for`, `foreach`, and `using` statement, creates a local variable declaration space for local variables and local constants. Names are introduced into this declaration space through *local_variable_declaration*s and *local_constant_declaration*s. Note that blocks that occur as or within the body of a function member or anonymous function are nested w ...
+- Each method declaration, property declaration, property accessor declaration, indexer declaration, indexer accessor declaration, operator declaration, instance constructor declaration, anonymous function, and local function creates a new declaration space called a ***local variable declaration space***. Names are introduced into this declaration space through formal parameters (*fixed_parameter ...
```
This code appears to break this restriction, for both a parameter and a local variable declared in a local function. (Ignore values etc - I adapted a piece of code I happened to be looking at.)

```csharp
using System;

class Test
{
    static void Main()
    {
        var f = float.MaxValue;
        var d = 10;

        Method("hello");
        Console.WriteLine(d+f);

        string Method(string f)
        {
            int d = 20;
            return d.ToString() + f;
        }

    }
}
```

jskeet replied:
This is covered in #590.

BillWagner replied:
WIth C# 7.3, the above program generates two errors:

```
1>error CS0136: A local or parameter named 'd' cannot be declared in this scope because that name is used in an enclosing local scope to define a local or parameter
1>error CS0136: A local or parameter named 'f' cannot be declared in this scope because that name is used in an enclosing local scope to define a local or parameter
```

In C# 8, and later, this compiles.

jskeet commented on standard/statements.md:
```diff
@@ -270,15 +270,18 @@ In addition to the reachability provided by normal flow of control, a labeled st
 ### 12.6.1 General
 
-A *declaration_statement* declares a local variable or constant. Declaration statements are permitted in blocks, but are not permitted as embedded statements.
+A *declaration_statement* declares a local variable, local constant, or local function. Declaration statements are permitted in blocks, but are not permitted as embedded statements.
```
I can't remember offhand what an embedded statement is, which means another reader might not either - maybe link to it?

jskeet replied:
Ah, it's saying that a declaration statement doesn't count as an embedded statement. Makes more sense.

jskeet replied:
(Not a change for local functions, so resolving.)

gafter commented on standard/statements.md:
```diff
@@ -414,6 +417,78 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+A local function is declared at block scope, and that function may capture variables from the enclosing scope. Captured variables must be definitely assigned before read by the body of the local function in each call to the function. The compiler shall determine which variables are definitely assigned on return (§definite-assignment-rules-for-local-function).
+
+A local function may be called from a lexical point prior to its definition. However, it is a compile-time error for the function to be declared lexically prior to the declaration of a variable it wishes to capture.
```
Change "it wishes to capture" to "used in the local function".

gafter replied:
Having said that, it is already a restriction in the language that a variable cannot be used except after its point of declaration.

BillWagner replied:
That restriction is in 7.7.

BillWagner replied:
Edited per https://github.com/dotnet/csharpstandard/pull/104#discussion_r965276993 and added a reference to §7.7

ericlippert commented on standard/variables.md:
```diff
@@ -802,6 +802,63 @@ For a *lambda_expression* or *anonymous_method_expression* *expr* with a body (e
+
+Local functions are analyzed in the context of their parent method. There are two control flow paths that matter for local functions: function calls and delegate conversions.
+
+Definite assignment for the body of each local function is defined separately for each call site. At each invocation, variables captured by the local function are considered definitely assigned if they were definitely assigned at the point of call. A control flow path also exists to the local function body at this point and is considered reachable. After a call to the local function, captured var ...
```
await statements, or expressions?

BillWagner replied:
should be "await expressions"

jskeet commented on standard/statements.md:
````diff
@@ -414,6 +417,78 @@ The scope of a local constant is the block in which the declaration occurs. It i
+    ;
+```
+
+Grammar note: When recognising a *local_function_body* if both the *null_conditional_invocation_expression* and *expression* alternatives are applicable then the former shall be chosen.
````
Can we reorder the options in the grammar then? In particular it's odd for the former in the prose to be the latter in the grammar, as well as making more sense (to me anyway) as "first match wins".

jskeet replied:
https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/classes.md#146-methods has null_conditional_invocation_expression first with the same text as is here.

But there's no example there either.

Suggest we add an example in 14.6 and refer to it from here.

jskeet replied:
11.7.9 gives a little more motivation, but still doesn't have an expression-bodied member example.

BillWagner replied:
This discussion is captured in #621 

Edit here to add a link to §14.6.1.

ericlippert commented on standard/statements.md:
```diff
@@ -414,6 +417,78 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+A *local_function_declaration* may include one `async` ([§14.15](classes.md#1415-async-functions)) modifier and one `unsafe` ([§22.1](unsafe-code.md#221-general)) modifier. If the declaration includes the `async` modifier then the return type shall be `void` or a task type  ([§14.15.1](classes.md#14151-general)). The `unsafe` modifier uses the containing lexical scope. The `async` modifier does n ...
+
+A local function is declared at block scope, and that function may capture variables from the enclosing scope. Captured variables must be definitely assigned before read by the body of the local function in each call to the function. The compiler shall determine which variables are definitely assigned on return (§definite-assignment-rules-for-local-function).
```
Is "before read" grammatical?

BillWagner replied:
reworded the sentence.

jskeet commented on standard/statements.md:
```diff
@@ -414,6 +417,78 @@ The scope of a local constant is the block in which the declaration occurs. It i
+>     }
+>     return alphabetSubsetImplementation();
+>
+>     IEnumerable<char> alphabetSubsetImplementation()
```
I normally write local functions to follow normal .NET method naming conventions, so this would be AlphabetSubsetImplementation. What do the MS docs normally do?

jskeet replied:
https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/local-functions is inconsistent - `nthFactorial` and `GetOddSequenceEnumerator` are both used as examples...

BillWagner replied:
I'd prefer `camelCaseLocalFunctionNames()`, because they are local in scope.  If you agree 👍 this comment. If you'd prefer `PascalCaseLocalFunctionsNames()`, because they are methods, 👎 this comment.

BillWagner replied:
Resolution from the committee: Let's use `PascalCaseLocalFunctionNames()`

jskeet commented on standard/statements.md:
```diff
@@ -414,6 +417,78 @@ The scope of a local constant is the block in which the declaration occurs. It i
+>
+>     IEnumerable<char> alphabetSubsetImplementation()
+>     {
+>         for (var c = start; c < end; c++)
```
Let's add braces around this :)

jskeet commented on standard/statements.md:
```diff
@@ -414,6 +417,78 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+Unless specified otherwise below, the semantics of all grammar elements is the same as for *method_declaration* ([§14.6.1](classes.md#1461-general)), read in the context of a local function instead of a method.
+
+The *identifier* of a *local_function_declaration* must be unique in its declared block scope. In other words, overloaded *local_function_declaration*s are not allowed.
```
And it can't be the same as any variables declared in the scope either?

jskeet replied:
Except that in C# 8 (at least) it's valid to declare a local variable within a local function, where the name of the local variable is also the name of a local function in the enclosing scope.

BillWagner replied:
The second sentence has been reworded. Instead of "In other words", it says "One consequence of this is that..."

Because variables are *identifiers*, the text is valid for C# 7 and later.

jskeet commented on standard/statements.md:
```diff
@@ -414,6 +417,78 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+The *identifier* of a *local_function_declaration* must be unique in its declared block scope. In other words, overloaded *local_function_declaration*s are not allowed.
+
+A *local_function_declaration* may include one `async` ([§14.15](classes.md#1415-async-functions)) modifier and one `unsafe` ([§22.1](unsafe-code.md#221-general)) modifier. If the declaration includes the `async` modifier then the return type shall be `void` or a task type  ([§14.15.1](classes.md#14151-general)). The `unsafe` modifier uses the containing lexical scope. The `async` modifier does n ...
```
Nit: double space after "a task type"

gafter commented on standard/statements.md:
```diff
@@ -414,6 +417,78 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+It is a compile-time error for a local function to declare a parameter or local variable with the same name as one declared in the enclosing scope.
+
+Local function bodies are always reachable. The endpoint of a local function is reachable if the beginning point of the local function is reachable.
```
```suggestion
Local function bodies are always reachable. The endpoint of a local function declaration is reachable if the beginning point of the local function declaration is reachable.
```

jskeet commented on standard/statements.md:
```diff
@@ -414,6 +417,78 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+A *local_function_declaration* may include one `async` ([§14.15](classes.md#1415-async-functions)) modifier and one `unsafe` ([§22.1](unsafe-code.md#221-general)) modifier. If the declaration includes the `async` modifier then the return type shall be `void` or a task type  ([§14.15.1](classes.md#14151-general)). The `unsafe` modifier uses the containing lexical scope. The `async` modifier does n ...
+
+A local function is declared at block scope, and that function may capture variables from the enclosing scope. Captured variables must be definitely assigned before read by the body of the local function in each call to the function. The compiler shall determine which variables are definitely assigned on return (§definite-assignment-rules-for-local-function).
```
"before read" => "before being read"?

jskeet replied:
```suggestion
A local function is declared at block scope, and that function may capture variables from the enclosing scope. Captured variables must be definitely assigned before being read by the body of the local function in each call to the function. The compiler shall determine which variables are definitely assigned on return (§definite-assignment-rules-for-local-function).
```

ericlippert commented on standard/statements.md:
````diff
@@ -414,6 +417,78 @@ The scope of a local constant is the block in which the declaration occurs. It i
+    ;
+```
+
+Grammar note: When recognising a *local_function_body* if both the *null_conditional_invocation_expression* and *expression* alternatives are applicable then the former shall be chosen.
````
This might be more clear with an example.

BillWagner replied:
See #621

gafter commented on standard/statements.md:
```diff
@@ -414,6 +417,78 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+Local function bodies are always reachable. The endpoint of a local function is reachable if the beginning point of the local function is reachable.
+
+Calls to local functions must be statically resolved.
```
Not sure what this means.  Should this be something said in the section describing the invocation expression, or the simple name expression?

gafter replied:
Perhaps would be more clear to say it may not be dynamically resolved?

jskeet replied:
We have two examples of what it prohibits:

- Use of `dynamic` for `params` which may change whether it's expanded or not
- Use of `dynamic` for type inference

(There may be others.)

BillWagner replied:
See Andy's second bullet in https://github.com/dotnet/csharpstandard/pull/104#issuecomment-1150455447.

If the argument to a local function is dynamic, the function to be called must be resolved at compile time, not runtime.

BillWagner replied:
Replaced with the following text:

> If the argument to a local function is dynamic, the function to be called must be resolved at compile time, not runtime.

jskeet commented on standard/statements.md:
```diff
@@ -414,6 +417,78 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+It is a compile-time error for a local function to declare a parameter or local variable with the same name as one declared in the enclosing scope.
+
+Local function bodies are always reachable. The endpoint of a local function is reachable if the beginning point of the local function is reachable.
```
What's the difference between the local function body being reachable and the beginning point of the local function being reachable? (I suspect I know, but this seems unnecessarily cryptic.)

BillWagner replied:
The discussion is here: https://github.com/dotnet/csharpstandard/pull/104#discussion_r616112191

I believe it is to have the correct rules on the reachability of any statement following a local function body. Maybe @gafter can suggest more clear language?

KalleOlaviNiemitalo replied:
An example could help

```C#
class C {
    int M() {
        // L();
        return 1;
        int L() { return 2; }
        return 3;
    }
}
```

Because of the statement `return 1;`, the beginning point of the local function L is not reachable. Consequently, the ending point of L is not reachable, and the statement `return 3;` is not reachable. The statement `return 2;` in the body of L is considered reachable even if L is not called.

jskeet commented on standard/statements.md:
```diff
@@ -414,6 +417,78 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+Local function bodies are always reachable. The endpoint of a local function is reachable if the beginning point of the local function is reachable.
+
+Calls to local functions must be statically resolved.
```
Is this all about prohibiting type inference using dynamic types? Or is there more to it?

gafter commented on standard/statements.md:
```diff
@@ -414,6 +417,78 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+A *local_function_declaration* may include one `async` ([§14.15](classes.md#1415-async-functions)) modifier and one `unsafe` ([§22.1](unsafe-code.md#221-general)) modifier. If the declaration includes the `async` modifier then the return type shall be `void` or a task type  ([§14.15.1](classes.md#14151-general)). The `unsafe` modifier uses the containing lexical scope. The `async` modifier does n ...
+
+A local function is declared at block scope, and that function may capture variables from the enclosing scope. Captured variables must be definitely assigned before read by the body of the local function in each call to the function. The compiler shall determine which variables are definitely assigned on return (§definite-assignment-rules-for-local-function).
```
"Captured" is not defined.

BillWagner replied:
See #622

jskeet commented on standard/unsafe-code.md:
```diff
@@ -18,11 +18,12 @@ An implementation that does not support unsafe code is required to diagnose any
 - A declaration of a class, struct, interface, or delegate may include an `unsafe` modifier, in which case, the entire textual extent of that type declaration (including the body of the class, struct, or interface) is considered an unsafe context.
   > *Note*: If the *type_declaration* is partial, only that part is an unsafe context. *end note*
 - A declaration of a field, method, property, event, indexer, operator, instance constructor, finalizer, or static constructor may include an `unsafe` modifier, in which case, the entire textual extent of that member declaration is considered an unsafe context.
+- A declaration of a local function may include an `unsafe` modifier, in which case, the entire textual extent of that local function is considered an unsafe context.
```
Any reason not to just include "local function" in the list of declarations above?

BillWagner replied:
I wrote this bullet to match the existing bullet items.

gafter commented on standard/variables.md:
```diff
@@ -802,6 +802,63 @@ For a *lambda_expression* or *anonymous_method_expression* *expr* with a body (e
+
+Local functions are analyzed in the context of their parent method. There are two control flow paths that matter for local functions: function calls and delegate conversions.
+
+Definite assignment for the body of each local function is defined separately for each call site. At each invocation, variables captured by the local function are considered definitely assigned if they were definitely assigned at the point of call. A control flow path also exists to the local function body at this point and is considered reachable. After a call to the local function, captured var ...
```
"captured" is not defined.

jskeet replied:
It's defined for anonymous function expressions in https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#111662-captured-outer-variables.

We could potentially leave it here without definition, but raise an issue to extract the concept of a captured variable to a common place and reference it from 11.16.6.2 and here.

BillWagner replied:
See #622

No change at this time. This should be addressed as part of #622

jskeet commented on standard/statements.md:
```diff
@@ -414,6 +417,78 @@ The scope of a local constant is the block in which the declaration occurs. It i
+>
+>     IEnumerable<char> alphabetSubsetImplementation()
+>     {
+>         for (var c = start; c < end; c++)
+>             yield return c;
```
```suggestion
>         for (var c = start; c < end; c++)
>         {
>             yield return c;
>         }
```

BillWagner committed:
Update standard/statements.md

Co-authored-by: Jon Skeet <jonskeet@google.com>

jskeet committed:
Update standard/statements.md

jskeet committed:
Update standard/statements.md

Co-authored-by: Neal Gafter <neal@gafter.com>

jskeet commented on standard/statements.md:
```diff
@@ -414,6 +417,80 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+The *identifier* of a *local_function_declaration* must be unique in its declared block scope. In other words, overloaded *local_function_declaration*s are not allowed.
+
+A *local_function_declaration* may include one `async` ([§14.15](classes.md#1415-async-functions)) modifier and one `unsafe` ([§22.1](unsafe-code.md#221-general)) modifier. If the declaration includes the `async` modifier then the return type shall be `void` or a task type  ([§14.15.1](classes.md#14151-general)). The `unsafe` modifier uses the containing lexical scope. The `async` modifier does n ...
```
```suggestion
A *local_function_declaration* may include one `async` ([§14.15](classes.md#1415-async-functions)) modifier and one `unsafe` ([§22.1](unsafe-code.md#221-general)) modifier. If the declaration includes the `async` modifier then the return type shall be `void` or a task type ([§14.15.1](classes.md#14151-general)). The `unsafe` modifier uses the containing lexical scope. The `async` modifier does not use the containing lexical scope. It is a compile-time error for *type_parameter_list* or *formal_parameter_list* to contain *attributes*.
```

jskeet committed:
Update standard/statements.md

jskeet commented on standard/statements.md:
```diff
@@ -414,6 +417,80 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+Unless specified otherwise below, the semantics of all grammar elements is the same as for *method_declaration* ([§14.6.1](classes.md#1461-general)), read in the context of a local function instead of a method.
+
+The *identifier* of a *local_function_declaration* must be unique in its declared block scope. In other words, overloaded *local_function_declaration*s are not allowed.
```
```suggestion
The *identifier* of a *local_function_declaration* must be unique in its declared block scope. One consequence of this is that overloaded *local_function_declaration*s are not allowed.
```

jskeet committed:
Update standard/statements.md

BillWagner committed:
edits based on September committee meeting

BillWagner commented:
@jskeet 

I've updated this PR based on all comments from the last meeting. There are two open conversations to resolve. The first, we need to pick a consensus. The second, we need to agree on the wording.

Can we add this to the next meeting agenda?

jskeet commented:
@BillWagner: Will do :)

BillWagner committed:
fix link text

BillWagner committed:
fix section renumbering issue

BillWagner committed:
Final edits

We resolved these two discussions during the committee meeting. Edits reflect those decisions.

BillWagner commented:
@jskeet Check the final commit, and when you're ready, let's :shipit:

jskeet commented on standard/statements.md:
```diff
@@ -414,6 +417,100 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+A *local_function_declaration* may include one `async` ([§14.15](classes.md#1415-async-functions)) modifier and one `unsafe` ([§22.1](unsafe-code.md#221-general)) modifier. If the declaration includes the `async` modifier then the return type shall be `void` or a task type ([§14.15.1](classes.md#14151-general)). The `unsafe` modifier uses the containing lexical scope. The `async` modifier does no ...
+
+A local function is declared at block scope, and that function may capture variables from the enclosing scope. It is a compile time error if a captured variables is read by the body of the local function if it is not definitely assigned before each call to the function. The compiler shall determine which variables are definitely assigned on return (§definite-assignment-rules-for-local-function).
```
Nit: "a captured variables" => "a captured variable".

jskeet replied:
And possibly change "if it is not" to "but is not". (It's a little odd to have the two "ifs" in the same sentence.) That could be a later change.

jskeet commented on standard/statements.md:
```diff
@@ -414,6 +417,100 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+A local function is declared at block scope, and that function may capture variables from the enclosing scope. It is a compile time error if a captured variables is read by the body of the local function if it is not definitely assigned before each call to the function. The compiler shall determine which variables are definitely assigned on return (§definite-assignment-rules-for-local-function).
+
+A local function may be called from a lexical point prior to its definition. However, it is a compile-time error for the function to be declared lexically prior to the declaration of a variable used in the local function (§7.7).
```
definition => declaration? (We have local function declarations, not local function definitions.)

jskeet commented on standard/statements.md:
```diff
@@ -414,6 +417,100 @@ The scope of a local constant is the block in which the declaration occurs. It i
+
+A *local_function_declaration* may include one `async` ([§14.15](classes.md#1415-async-functions)) modifier and one `unsafe` ([§22.1](unsafe-code.md#221-general)) modifier. If the declaration includes the `async` modifier then the return type shall be `void` or a task type ([§14.15.1](classes.md#14151-general)). The `unsafe` modifier uses the containing lexical scope. The `async` modifier does no ...
+
+A local function is declared at block scope, and that function may capture variables from the enclosing scope. It is a compile time error if a captured variables is read by the body of the local function if it is not definitely assigned before each call to the function. The compiler shall determine which variables are definitely assigned on return (§definite-assignment-rules-for-local-function).
```
"compile time" => "compile-time"

jskeet reviewed (approved):
Just a few nits left; happy for it to be merged after that.

BillWagner committed:
updates from code review.

BillWagner committed:
one more bit of feedback.
